ndonnx/_typed_array/onnx.py

```diff
@@ -0,0 +1,2125 @@
+        self, /, *, axis: int | tuple[int, ...] | None = None, keepdims: bool = False
+    ) -> TyArrayBool:
+        if axis == ():
+            return self.astype(bool_)
+        if isinstance(axis, int):
+            axis = (axis,)
+
+        bools = self.astype(bool_)
+
+        if bools.ndim == 0:
+            if axis:
+                ValueError("'axis' were provided but 'self' is a scalar")
+            # Nothing left to reduce
+            return safe_cast(TyArrayBool, bools)
+
+        axes = op.const(list(axis), np.int64) if axis else None
+
+        # Accumulate in float32 to avoid possible overflowing issues
+        var = op.reduce_sum(bools.astype(float32)._var, axes=axes, keepdims=keepdims)
+        return safe_cast(TyArrayBool, TyArrayFloat32(var).astype(bool_))
+
+    def as_core_dtype(self, dtype: DTypes) -> TyArray:
+        raise ValueError(f"Casting between `{self.dtype}` and `{dtype}` is undefined")
+
+    def broadcast_to(self, shape: tuple[int, ...] | TyArrayInt64) -> Self:
+        if isinstance(shape, tuple):
+            shape = astyarray(np.array(shape), dtype=int64)
+        if shape.ndim != 1:
+            raise ValueError(f"'shape' must be a rank-1 tensor, found `{shape.ndim}`")
+        if not isinstance(shape.shape[0], int):
+            raise ValueError("'shape' must be a rank-1 tensor with static length.")
+        (target_rank,) = shape.shape
+        if not isinstance(target_rank, int):
+            raise ValueError(
+                "target rank of broadcasting operation must be statically known"
+            )
+        if target_rank < self.ndim:
+            raise ValueError("target rank must equal or greater than rank of 'self'")
+        var = op.expand(self._var, shape._var)
+        return type(self)(var)
+
+    def concat(self, others: Sequence[Self], axis: None | int) -> Self:
+        others = list(others)
+        arrays = [self] + others
+        if axis is None:
+            arrays = [a if a.ndim == 1 else a.reshape((-1,)) for a in arrays]
+            axis = 0
+
+        seen_ranks = {a.ndim for a in arrays}
+        if len(seen_ranks) != 1:
+            raise ValueError("concatenated arrays must all have the same rank")
+        (rank,) = seen_ranks
+        if rank == 0:
+            raise ValueError("zero-dimensional arrays cannot be concatenated")
+
+        if len(arrays) == 1:
+            return arrays[0].copy()
+
+        # It seems that there is currently a bug(?) in the type/shape
+        # inference in ONNX which prohibits us from concatenating
+        # empty 1D int32 and int64 arrays (see test case). We therefor
+        # do some hacky special-casing here for those types and those
+        # types only. Other data types are fine.
+        #
+        # TODO: File upstream bug; this may also be what caused the
+        # segfaults in onnxruntime in the past!
+        if self.dtype in (int32, int64) and self.ndim == 1:
+            axis_ = axis if axis is None else axis
```

Comment:
Why is this line producing `axis` regardless of branch.

Reply:
Well spotted!